### PR TITLE
Better error message for 'class func/var' usage in protocols

### PR DESCRIPTION
**Pull request by Moximillian**
Clarify how to use protocols in relation to class funcs. This has been discussed in https://forums.swift.org/t/class-constrained-protocols-dont-allow-for-class-methods/13031

### CODE REVIEW

**Comment by xwu** [comment]
I don't think this is an appropriate fix when the usage being diagnosed is in an `enum`, for example. It should be restricted to usage in protocol declarations only.

Consider perhaps also reworking the wording to omit unnecessary text (such as "keyword") and to clarify what's actually being advised (e.g., "use 'static' to declare a requirement fulfilled by either a class or static method").

File: include/swift/AST/DiagnosticsCommon.def [hunk 1]
```diff
@@ -66,10 +66,10 @@ ERROR(super_not_in_class_method,none,
 
 ERROR(class_func_not_in_class,none,
       "class methods are only allowed within classes; "
-      "use 'static' to declare a static method", ())
+      "use 'static' to declare a requirement fulfilled by either a class or static method", ())
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsCommon.def]: "requirement" is the language one uses in the context of protocol members.  I think the confusion in the thread stems from the `class` requirement on protocols not emitting a more refined diagnostic.  You should look into specializing this for exactly that case.

**Review by CodaFi — CHANGES_REQUESTED** [review]

**Comment by Moximillian** [comment]
@xwu updated the error text as a separate error, so that protocols can have different text than structs and enums. Only protocol related tests have changes, other tests are unchanged.
Tried to apply the protocol specific error in ParseDecl.cpp. Does TypeCheckDecl.cpp need to be changed too?

**Comment by CodaFi** [comment]
I still think this isn't *quite* what the original thread intended.  OP seemed confused that a `class` bound on a protocol would emit a diagnostic that, the right light, implies classes wouldn't be able to conform to the protocol because `class` members aren't `static` members.  I don't think the original diagnostics need updating, I think we need a new diagnostic to address that head on.

**Comment by Moximillian** [comment]
Apparently extension parsing/type checking happens in TypeCheckDecl.cpp, so for protocol extensions, the appropriate error is not yet produced.

**Comment by CodaFi** [comment]
```swift
protocol MyProtocol: class {
    static func a() -> Self // Compiles.
    class func b() -> Self // Doesn't compile.  "Class methods are only allowed within classes; use 'static' to declare a static method"
}
```

This is the original.  It isn't about the conformances, it's about the protocol requirements.

**Comment by CodaFi** [comment]
>  It's not obvious for new users.

Okay, I'll re-review under this new specification.

File: include/swift/AST/DiagnosticsCommon.def [hunk 2]
```diff
@@ -71,6 +71,13 @@ ERROR(class_var_not_in_class,none,
       "class properties are only allowed within classes; "
       "use 'static' to declare a static property", ())
 
+ERROR(class_func_not_in_class_protocol,none,
+      "class methods are only allowed within classes; "
+      "use 'static' to declare a requirement fulfilled by either a class or static method", ())
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsCommon.def]: When diagnostics are related by only a snippet of text, as this one is to its neighbors to the north, you should parametrize the diagnostic.  [For example](https://github.com/apple/swift/blob/master/include/swift/AST/DiagnosticsSema.def#L305-L306)

File: lib/Parse/ParseDecl.cpp [hunk 3]
```diff
@@ -4808,7 +4808,7 @@ Parser::parseDeclVar(ParseDeclOptions Flags,
     } else if (Flags.contains(PD_InStruct) || Flags.contains(PD_InEnum) ||
                Flags.contains(PD_InProtocol)) {
       if (StaticSpelling == StaticSpellingKind::KeywordClass)
-        diagnose(Tok, diag::class_var_not_in_class)
+        diagnose(Tok, Flags.contains(PD_InProtocol) ? diag::class_var_not_in_class_protocol : diag::class_var_not_in_class)
```
Comment (CodaFi) [inline lib/Parse/ParseDecl.cpp]: Nit: This line should be wrapped to 80 characters.  You can pull out diagnostic selectors as variables or parametrize the diagnostic.  Either one would shorten the line.

File: lib/Parse/ParseDecl.cpp [hunk 4]
```diff
@@ -5194,7 +5194,7 @@ Parser::parseDeclFunc(SourceLoc StaticLoc, StaticSpellingKind StaticSpelling,
     } else if (Flags.contains(PD_InStruct) || Flags.contains(PD_InEnum) ||
                Flags.contains(PD_InProtocol)) {
       if (StaticSpelling == StaticSpellingKind::KeywordClass) {
-        diagnose(Tok, diag::class_func_not_in_class)
+        diagnose(Tok, Flags.contains(PD_InProtocol) ? diag::class_func_not_in_class_protocol : diag::class_func_not_in_class)
```
Comment (CodaFi) [inline lib/Parse/ParseDecl.cpp]: Nit: ditto.

File: include/swift/AST/DiagnosticsCommon.def [hunk 5]
```diff
@@ -71,6 +71,13 @@ ERROR(class_var_not_in_class,none,
...
+ERROR(class_var_not_in_class_protocol,none,
+      "class properties are only allowed within classes; "
+      "use 'static' to declare a requirement fulfilled by either a class or static property", ())
```
[the remark this answers is not in the document]
Comment (xwu) [reply include/swift/AST/DiagnosticsCommon.def]: The last diagnostic in this file begins: “Generic requirement...” I doesn’t strike me as unfriendly.

**Comment by Moximillian** [comment]
@CodaFi thanks for the review, I'll work on this...

File: include/swift/AST/DiagnosticsCommon.def (hunk 5 above)
[the remark this answers is not in the document]
Comment (CodaFi) [reply include/swift/AST/DiagnosticsCommon.def]: Whoops, grep'ed too hard in the wrong directory.  Striking this.

**Comment by CodaFi** [comment]
Dropped my comments about the wording so we're just down to the nits then, @Moximillian.  I'll take this once those are resolved.

File: include/swift/AST/DiagnosticsCommon.def (hunk 5 above)
[the remark this answers is not in the document]
Comment (Moximillian) [reply include/swift/AST/DiagnosticsCommon.def]: Just for the sake of completeness:
**Protocols**
“A protocol defines a blueprint of methods, properties, and other *requirements* that suit a particular task or piece of functionality. ”

Excerpt From: Apple Inc. “The Swift Programming Language (Swift 4).” iBooks.

**Comment by Moximillian** [comment]
@xwu @CodaFi parametrized the error, using 80 columns for the code and implemented error for the protocol extensions too (& tests for it)

**Comment by xwu** [comment]
The error message doesn’t make sense to me for protocol extensions: you cannot declare requirements in an extension, only statically dispatched protocol extension methods, so the original error message is correct.

**Comment by Moximillian** [comment]
@xwu you are correct. Updated the protocol extensions to use the original version of the error.

**Comment by xwu** [comment]
@swift-ci Please smoke test

**Comment by Moximillian** [comment]
Test failures have been fixed

**Comment by xwu** [comment]
@swift-ci Please smoke test

**Review by CodaFi — APPROVED** [review]

**Comment by CodaFi** [comment]
⛵️

**Comment by AnthonyLatsis** [comment]
Leaving out the error message as a whole, I am still not happy with the `fulfilled by either a class or static method`. It's read as if the "or" is between a class and a static method rather than between a class method or a static method.

**Comment by xwu** [comment]
That reading is syntactically possible but semantically absurd; I suppose we could reverse the order and write "static or class method" but I'm doubtful this is a problem. I'm not familiar with the phrase "type method" in the context of Swift diagnostics.

**Comment by Moximillian** [comment]
I tend to agree with switching the static/class in the error (even if it's just so tiny change). Created a new PR for it.

**Pull request merged**